util/tune/current_settings.rb

```diff
@@ -0,0 +1,164 @@
+# frozen_string_literal: true
+
+require "json"
+
+NA = "N/A"
+PE_PUPPET_SERVER_CONF = "/etc/puppetlabs/puppetserver/conf.d/pe-puppet-server.conf"
+POSTGRES_CONF = "/opt/puppetlabs/server/data/postgresql/9.6/data/postgresql.conf"
+PUPPET_DB_CONF = "/etc/puppetlabs/puppetdb/conf.d/config.ini"
+
+def puppetserver_jruby_max_active_instances
+  if File.exist? PE_PUPPET_SERVER_CONF
+    command = "cat #{PE_PUPPET_SERVER_CONF} | grep max-active-instances | egrep -v '^#' | cut -d ':' -f 2"
+    output = `#{command}`.strip!
+    value = output.to_i unless output.nil? || output.empty?
+  end
+
+  unless value
+    command = "facter processorcount"
+    output = `#{command}`
+
+    if output
+      num_cores = output.to_i
+      minimum = 1
+      maximum = 4
+      value = [[num_cores, minimum].max, maximum].min
+    else
+      value = NA
+    end
+
+  end
+
+  value
+end
+
+def get_java_args(file)
+  path = if File.exist? "/etc/debian_version"
+           "/etc/defaults/#{file}"
+         else
+           "/etc/sysconfig/#{file}"
+         end
+
+  if File.exist? path
+    command = "cat #{path} | grep Xmx | grep Xmx"
+    output = `#{command}`
+    value = output.split('"')[1] if output
+  else
+    value = NA
+  end
+
+  value
+end
+
+def puppetserver_reserved_code_cache
+  path = if File.exist? "/etc/debian_version"
+           "/etc/defaults/pe-puppetserver"
+         else
+           "/etc/sysconfig/pe-puppetserver"
+         end
+
+  if File.exist? path
+    command = "cat #{path} | grep Xmx | grep Xmx"
+    res_param = nil
+
+    output = `#{command}`
+    output.split(" ").each do |param|
+      if param.include? "ReservedCodeCacheSize"
+        res_param = param
+        break
+      end
+    end
+
+    value = res_param.split("=")[1] if res_param
+  else
+    value = NA
+  end
+
+  value
+end
+
+def console_java_args
+  get_java_args("pe-console-services")
+end
+
+def master_java_args
+  get_java_args("pe-puppetserver")
+end
+
+def orchestrator_java_args
+  get_java_args("pe-orchestration-services")
+end
+
+def puppetdb_java_args
+  get_java_args("pe-puppetdb")
+end
+
+# TODO: add exclude arg
+def get_postgres_parameter(parameter)
+  command = "cat #{POSTGRES_CONF} | grep #{parameter} | egrep -v '^#' | cut -d '=' -f 2 | cut -d '#' -f 1"
```

Comment:
More bash, hrm.  Since we are in ruby, it seems like it would be more efficient to parse the config file into a hash and then lookup values as necessary rather than running this bash command each time. Possible example:
```
def parse_psql_conf()
  psql_conf = {}
  File.open(POSTGRES_CONF).each do |line|
    if line.strip.split('#').first =~ /\=/
      key, value = line.strip.split('#').first.split('=')
      key = key.strip.to_sym
      value.gsub!(/\A"|"\Z/, '')
      value = value.strip
      psql_conf[key] = value
    end
  end
end

psql_conf = parse_psql_conf
psql_conf[:shared_buffers]
...
```
This is a rough sketch. Let me know if the idea is unclear.

Reply:
I have updated the code to eliminate wrapping bash in ruby as noted above and abstracted it to use a `get_conf_parameter` method called by the existing `get_postgres_parameter` and a new `get_puppetdb_parameter` which is used by `puppetdb_command_processing_threads`.

If the efficiency of opening the file for each database setting is a bottleneck I would agree with the approach you suggested. However, in a quick test using Bolt to run the script, the fastest time I was able to achieve checking a single database setting (all other lines commented out) was 2.79 seconds (5 runs). Checking all of the settings (no lines commented) was 2.84 seconds.